// @node-red/nodes/core/timenode/time.js
module.exports = function (RED) {
    function TimeNode(config) {
        RED.nodes.createNode(this, config);
        var node = this;

        // 获取配置信息
        this.dateseparator = config.dateseparator || "-";
        this.timeseparator = config.timeseparator || "-";
        this.connect = config.connect || "-";
        this.dateselect = config.dateselect || "-";
        this.timeselect = config.timeselect || "-";
        // 处理输入消息
        this.on('input', function (msg) {
            // 获取当前时间
            var currentTime = new Date();

            // 格式化日期
            var formattedDate = formatDate(currentTime, node.dateselect,node.timeselect,node.dateseparator, node.timeseparator, node.connect);

            // 将格式化后的日期作为输出消息发送
            msg.payload = formattedDate;
            node.send(msg);
        });
    }

    // 格式化日期函数
    function formatDate(date, dateselect,timeselect,dateseparator, timeseparator, connect) {
        var year = date.getFullYear();
        var month = date.getMonth() + 1;
        var day = date.getDate();
        var week =date.getDay();
        switch (week) {
            case 7:
                week= "星期日";
            break;
            case "6":
                week= "星期六";
                break;
            case "5":
                week= "星期五";
                break;
            case "4":
                week= "星期四";
                break;
            case 3:
                week= "星期三";
                break;
            case "2":
                week= "星期二";
                break;
            case "1":
                week= "星期一";
                break;
        }
        var hour = date.getHours();
        var minute = date.getMinutes();
        var second = date.getSeconds();
        var formattedDate = "";
        var formattedTime = "";
        var newDate = "";
        // 根据配置的日期格式进行格式化
        switch (dateselect) {
            case "five":
                formattedDate+=year+dateseparator+month+dateseparator+day+dateseparator+week+connect;
                break;
            case "four":
                formattedDate+=week+dateseparator+month+dateseparator+day+dateseparator+year+connect;
                break;
            case "three":
                formattedDate+=year+dateseparator+month+dateseparator+day+connect;
                break;
            case "two":
                formattedDate+=month+dateseparator+day+connect;
                break;
            case "one":
                formattedDate+=day+connect;
                break;
            default:
                formattedDate = "";
                break;
        }
        switch (timeselect) {
            case "three":
                formattedTime += hour+timeseparator+minute+timeseparator+second;
                break;
            case "two":
                formattedTime += minute+timeseparator+second;
                break;
            case "one":
                formattedTime += second;
                break;
            default:
                formattedTime = "";
                break;
        }
        newDate +=formattedDate+formattedTime;
        //newDate+=date+ dateselect+timeselect+dateseparator+timeseparator+connect;
        return newDate;
    }

    // 注册节点类型
    RED.nodes.registerType('time', TimeNode);
};



//npm install E:\lsd学习文件\Node-Red\node-red\packages\node_modules\@node-red\nodes\core\mynode
